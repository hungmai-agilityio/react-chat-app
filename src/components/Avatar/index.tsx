import { clsx } from 'clsx';

// Constants
import { SIZE } from '@/constants';

interface AvatarProps {
  avatar?: string;
  name?: string;
  size?: SIZE.SMALL | SIZE.LARGE;
  styles?: string;
  circle?: boolean;
  onClick?: () => void;
}

const Avatar = ({
  avatar,
  name,
  size,
  styles,
  circle,
  onClick
}: AvatarProps) => {
  const nameAvatar = name ? name.charAt(0).toUpperCase() : '';

  const sizeClasses = {
    [SIZE.SMALL]: 'w-avatar-sm h-avatar-sm text-lg',
    [SIZE.LARGE]: 'w-avatar-lg h-avatar-lg text-8xl'
  };

  const sizeClass = sizeClasses[size || SIZE.SMALL];

  if (!avatar) {
    return (
      <div
        data-testid="avatar"
        onClick={onClick}
        className={clsx(
          'text-primary bg-gray-300 flex items-center justify-center font-semibold',
          circle && 'rounded-full',
          styles,
          sizeClass
        )}
      >
        {nameAvatar}
      </div>
    );
  }

  return (
    <img
      data-testid="avatar"
      onClick={onClick}
      src={avatar}
      alt={name}
      className={clsx(
        'object-cover font-bold',
        circle && 'rounded-full',
        styles,
        sizeClass
      )}
    />
  );
};

export default Avatar;
