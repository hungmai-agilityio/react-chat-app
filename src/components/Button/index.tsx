import { clsx } from 'clsx';
import { memo } from 'react';

// Constants
import { SIZE, TYPE } from '@/constants';

// FontAwesome
import { IconDefinition } from '@fortawesome/fontawesome-svg-core';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

interface ButtonProps {
  name: string;
  size?: SIZE.MINI | SIZE.SMALL | SIZE.MEDIUM | SIZE.LARGE;
  variant?: TYPE.PRIMARY | TYPE.SECOND | TYPE.TERTIARY;
  onClick: (e: React.FormEvent<HTMLButtonElement>) => void;
  styles?: string;
  isDisabled?: boolean;
  iconLeft?: IconDefinition;
  iconRight?: IconDefinition;
}

const Button = ({
  name,
  size,
  variant,
  onClick,
  styles,
  isDisabled,
  iconLeft,
  iconRight
}: ButtonProps) => {
  const sizeClasses = {
    [SIZE.MINI]: 'w-24 h-9',
    [SIZE.SMALL]: 'w-32 h-10',
    [SIZE.MEDIUM]: 'w-40 h-9',
    [SIZE.LARGE]: 'w-full h-10'
  };

  const typeClasses = {
    [TYPE.PRIMARY]: 'bg-primary text-white',
    [TYPE.SECOND]: 'bg-tertiary text-primary',
    [TYPE.TERTIARY]: 'hover:bg-tertiary text-primary'
  };

  const sizeClass = sizeClasses[size || SIZE.SMALL];
  const typeClass = typeClasses[variant || TYPE.PRIMARY];

  return (
    <button
      className={clsx(
        'font-semibold hover:brightness-90',
        isDisabled &&
          'bg-tertiary cursor-not-allowed border hover:filter-none text-stone-500',
        sizeClass,
        typeClass,
        styles
      )}
      onClick={onClick}
      disabled={isDisabled}
    >
      {iconLeft && <FontAwesomeIcon icon={iconLeft} className="mr-2" />}
      {name}
      {iconRight && <FontAwesomeIcon icon={iconRight} className="ml-2" />}
    </button>
  );
};

const MemoizedButton = memo(Button);
export default MemoizedButton;
