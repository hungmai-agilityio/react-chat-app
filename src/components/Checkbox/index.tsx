import { clsx } from 'clsx';
import { ReactNode } from 'react';

interface CheckboxProps {
  id: string;
  children: ReactNode;
  isChecked: boolean;
  onChange: (id: string) => void;
}

const Checkbox = ({ id, children, isChecked, onChange }: CheckboxProps) => {
  // Handle checkbox
  const handleChecked = () => {
    onChange(id);
  };

  return (
    <div
      id={id}
      className={clsx('pl-3 flex items-center gap-4 w-full', {
        'bg-blue-50': isChecked
      })}
    >
      <input
        type="checkbox"
        checked={isChecked}
        className="w-5 h-5 accent-primary"
        onChange={handleChecked}
      />
      <label onClick={handleChecked} className="w-full">
        {children}
      </label>
    </div>
  );
};

export default Checkbox;
