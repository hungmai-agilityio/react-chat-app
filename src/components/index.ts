// Exports default
export { default as Avatar } from '@/components/Avatar';
export { default as Button } from '@/components/Button';
export { default as Divider } from '@/components/Divider';
export { default as Dropdown } from '@/components/Dropdown';
export { default as UserMenu } from '@/components/Dropdown/UserMenu';
export { default as ErrorBoundary } from '@/components/ErrorBoundary';
export { default as Form } from '@/components/Form';
export { default as SignIn } from '@/components/Form/SignIn';
export { default as SignUp } from '@/components/Form/SignUp';
export { default as Input } from '@/components/Input';
export { default as Loading } from '@/components/Loading';
export { default as Spinner } from '@/components/Loading/Spinner';
export { default as ChatMessage } from '@/components/Chat';

export { default as Modal } from '@/components/Modal';
export { default as ModalAction } from '@/components/Modal/Action';
export { default as AddMember } from '@/components/Modal/AddMember';
export { default as ModalInfo } from '@/components/Modal/Info';
export { default as Profile } from '@/components/Modal/Profile';
export { default as Search } from '@/components/Search';
export { default as Upload } from '@/components/Upload';
export { default as ListUser } from '@/components/UserRoom/List';

// Exports named
export * from '@/components/Checkbox';
export * from '@/components/Dropdown';
export * from '@/components/Chip';
export * from '@/components/Item';
export * from '@/components/Message';
export * from '@/components/Message/GroupMessage';
export * from '@/components/Tabs';
export * from '@/components/Tabs/Tab';
export * from '@/components/Tabs/Panel';
export * from '@/components/UserRoom';
