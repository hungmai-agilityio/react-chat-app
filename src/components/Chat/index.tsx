import {
  ChangeEvent,
  Dispatch,
  SetStateAction,
  useCallback,
  useEffect,
  useRef,
  useState
} from 'react';
import { v4 as uuidv4 } from 'uuid';
import clsx from 'clsx';

// Hooks
import { useOutsideClick } from '@/hooks';

// Components
import { Button, Input } from '@/components';

// Interfaces
import { IMessage, IChat } from '@/interfaces';

// Services
import { createChat, sendMessage, updateMessage } from '@/services';

// Constants
import { SIZE, TYPE } from '@/constants';

interface ChatMessageProps {
  currentUserId: string;
  setMessages: Dispatch<SetStateAction<IMessage[]>>;
  isDisabled: boolean;
  chatData?: IChat | null;
  selectedUser?: string;
  chats?: IChat[];
  handleCancelEdit: () => void;
  isEdit: boolean;
  editMessage?: IMessage | null;
}

const ChatMessage = ({
  currentUserId,
  setMessages,
  isDisabled,
  chatData,
  selectedUser,
  chats,
  handleCancelEdit,
  isEdit,
  editMessage
}: ChatMessageProps) => {
  const [value, setValue] = useState<string>('');
  const [isFocused, setIsFocused] = useState<boolean>(false);

  const inputRef = useRef<HTMLDivElement>(null);

  // Hook to handle outside click
  useOutsideClick({
    ref: inputRef,
    handler: () => setIsFocused(false)
  });

  // Initialize value if editing
  useEffect(() => {
    if (isEdit && editMessage) {
      setValue(editMessage.message);
    }
  }, [isEdit, editMessage]);

  // Handle change value message
  const handleChangeValue = (
    event: ChangeEvent<HTMLTextAreaElement | HTMLInputElement>
  ) => {
    setValue(event.target.value);
  };

  // Handle send message
  const handleSend = useCallback(async () => {
    if (!value.trim().length) return;

    let roomId = chatData?.id;

    if (!roomId && selectedUser) {
      const existingChat = chats?.find(
        (chat) =>
          !chat.isGroup &&
          chat.members.includes(currentUserId) &&
          chat.members.includes(selectedUser)
      );

      if (!existingChat) {
        const newChat: IChat = {
          id: uuidv4(),
          title: '',
          avatar: '',
          members: [currentUserId, selectedUser],
          isGroup: false
        };

        const createChatResponse = await createChat(newChat);
        roomId = createChatResponse.data?.id;
      }

      if (existingChat) {
        roomId = existingChat?.id;
      }
    }

    const newMessage: IMessage = {
      id: uuidv4(),
      message: value,
      sender: currentUserId,
      time_stamp: new Date(),
      roomId: roomId || ''
    };

    await sendMessage(newMessage);

    setMessages((prevMessages: IMessage[]) => {
      const messageExists = prevMessages.some(
        (msg) => msg.id === newMessage.id
      );
      if (messageExists) return prevMessages;

      return [...prevMessages, newMessage];
    });
    setValue('');
  }, [value, currentUserId, chatData, selectedUser, chats]);

  // Handle update message
  const handleUpdateMessage = useCallback(async () => {
    if (!value.trim().length || !editMessage) return;

    const updatedMessage: Partial<IMessage> = {
      id: editMessage.id,
      message: value,
      sender: currentUserId,
      isEdit: true
    };

    const { data } = await updateMessage(updatedMessage as IMessage);

    if (data) {
      setMessages((prev) =>
        prev.map((msg) =>
          msg.id === editMessage.id
            ? { ...msg, message: value, isEdit: true }
            : msg
        )
      );
    }

    setValue('');
    handleCancelEdit();
  }, [value, editMessage, currentUserId, handleCancelEdit]);

  // Handle press enter to send message
  const handleKeyDown = (
    e: React.KeyboardEvent<HTMLTextAreaElement | HTMLInputElement>
  ) => {
    if (e.key === 'Enter' && !e.shiftKey) {
      e.preventDefault();
      if (isEdit) {
        handleUpdateMessage();
      } else {
        handleSend();
      }
    }
  };

  return (
    <div
      ref={inputRef}
      onFocus={() => setIsFocused(true)}
      className={clsx('relative p-2 transition-all')}
    >
      <Input
        htmlFor="message-value"
        onChange={handleChangeValue}
        name="message"
        value={value}
        variant={TYPE.SECOND}
        disabled={isDisabled}
        onKeyDown={handleKeyDown}
      />
      {isFocused && (
        <div className="float-right">
          <div className="flex gap-5">
            {isEdit && (
              <Button
                name="Cancel"
                onClick={handleCancelEdit}
                size={SIZE.MINI}
              />
            )}
            <Button
              name={isEdit ? 'Update' : 'Send'}
              onClick={isEdit ? handleUpdateMessage : handleSend}
              disabled={value.length === 0 || isDisabled}
              size={SIZE.MINI}
            />
          </div>
        </div>
      )}
    </div>
  );
};

export default ChatMessage;
