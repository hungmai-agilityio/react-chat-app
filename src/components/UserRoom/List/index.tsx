import { memo } from 'react';

// Interfaces
import { IChat, IMessage, IProfile, IUser } from '@/interfaces';

// Hooks
import { useUsersWithProfiles, useOnlineStatus } from '@/hooks';

// Components
import { UserProps, UserRoom } from '@/components/UserRoom';

interface ListProps extends UserProps {
  data: (IUser | IChat)[];
  profiles: Record<string, IProfile>;
  currentUser?: string;
  onSelected: (id: string, isUser: boolean) => void;
  messages?: IMessage[];
}

const ListUser = memo(
  ({ data, profiles, onSelected, currentUser, messages }: ListProps) => {
    const { users } = useUsersWithProfiles();
    const status = useOnlineStatus(currentUser);

    const getUserDetails = (item: IUser | IChat) => {
      const id = item.id;
      const isUser = 'userName' in item;
      const name = isUser ? item.userName : item.title;
      const avatar = isUser ? profiles[id]?.avatar : item.avatar;

      if (isUser || item.isGroup) {
        return { name, avatar, otherUserId: '' };
      }

      const otherMemberId = item.members.find(
        (member) => member !== currentUser
      );

      const otherUser = users?.find((user) => user.id === otherMemberId);

      if (!otherUser) {
        return { name, avatar, otherUserId: '' };
      }

      return {
        name: item.title || otherUser.userName,
        avatar: item.avatar || profiles[otherUser.id]?.avatar,
        otherUserId: otherUser.id
      };
    };

    return data.map((item) => {
      const { name, avatar, otherUserId } = getUserDetails(item);
      const id = item.id;
      const isUser = 'userName' in item;
      const lastMessage = messages?.find((msg) => msg.roomId === id);
      const isActive =
        status[id] || (otherUserId && status[otherUserId]) || false;

      return (
        <UserRoom
          key={id}
          id={id}
          isCurrentUser={item.id === currentUser}
          name={name}
          avatar={avatar}
          isActive={isActive}
          path={`/chat/${id}`}
          onSelect={onSelected.bind(null, id, isUser)}
          isGroup={!isUser && item.isGroup}
          message={lastMessage?.message}
          time_stamp={lastMessage?.time_stamp.toString()}
        />
      );
    });
  }
);

export default ListUser;
