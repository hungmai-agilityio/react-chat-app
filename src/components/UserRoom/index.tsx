import { memo } from 'react';
import { clsx } from 'clsx';
import { NavLink } from 'react-router-dom';

// Constants
import { SIZE } from '@/constants';

// FontAwesome
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faCircle } from '@fortawesome/free-solid-svg-icons';

// Utils
import { timeStampConversion } from '@/utils';

// Components
import Avatar from '@/components/Avatar';

export interface UserProps {
  id?: string;
  isCurrentUser?: boolean;
  name?: string;
  avatar?: string;
  isActive?: boolean;
  isGroup?: boolean;
  isGroupChat?: boolean;
  time_stamp?: string;
  message?: string;
  path?: string;
  onSelect?: () => void;
}

export const UserRoom = memo(
  ({
    id,
    isCurrentUser,
    name,
    avatar,
    isActive,
    isGroup,
    time_stamp,
    message,
    path,
    onSelect
  }: UserProps) => {
    return (
      <NavLink
        key={id}
        to={path!}
        className="flex justify-between items-center w-full h-16 text-sm hover:bg-second px-3"
        onClick={onSelect}
      >
        <div className="flex items-center gap-3">
          <div className="relative">
            <Avatar name={name} avatar={avatar} circle size={SIZE.SMALL} />
            {!isGroup && (
              <FontAwesomeIcon
                icon={faCircle}
                className={clsx(
                  'absolute text-gray-400 border-xs border-white rounded-full top-6 left-7',
                  { 'text-green-500': isActive },
                  { 'text-red-500': isCurrentUser }
                )}
              />
            )}
          </div>
          {message ? (
            <div>
              <p className="font-semibold text-sm w-36 truncate">
                {isGroup ? `${name}` : `PM: ${name}`}
              </p>
              <p className="w-36 truncate">{message || ''}</p>
            </div>
          ) : (
            <p>{name}</p>
          )}
        </div>
        <p>{time_stamp && timeStampConversion(time_stamp)}</p>
      </NavLink>
    );
  }
);
