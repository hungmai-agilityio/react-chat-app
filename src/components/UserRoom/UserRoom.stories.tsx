import { Meta, StoryObj } from '@storybook/react';
import { withRouter } from 'storybook-addon-remix-react-router';

// Components
import { UserRoom } from '@/components/UserRoom';

// Components

const meta: Meta<typeof UserRoom> = {
  component: UserRoom,
  decorators: [withRouter],
  tags: ['autodocs'],
  argTypes: {
    id: {
      description: 'Id user'
    },
    name: {
      description: 'user name or chat room name'
    },
    avatar: {
      description: 'User avatar or chat room avatar'
    },
    isActive: {
      description: 'Check user active and render status'
    },
    isGroup: {
      description: 'Check chat room is member or group chat'
    },
    time_stamp: {
      description: 'render the time the last message was sent'
    },
    message: {
      description: 'render the last message was sent'
    },
    path: {
      description:
        'The link opens the chat area according to the user id or chat room'
    },
    onSelect: {
      description: 'Select chat room or user by id to start a conversation'
    }
  }
};

export default meta;

type Story = StoryObj<typeof UserRoom>;

export const GroupChat: Story = {
  args: {
    name: 'User',
    avatar: '',
    isActive: true,
    isGroup: true,
    message: 'Hello',
    time_stamp: '2024-05-24T15:58:05.791Z',
    path: '/'
  }
};

export const ChatMember: Story = {
  args: {
    name: 'User',
    avatar: 'assets/images/mock-image.jpg',
    isActive: true,
    message: 'Hello you',
    time_stamp: new Date().toISOString(),
    path: '/'
  }
};

export const UserOnline: Story = {
  args: {
    name: 'User',
    avatar: '',
    isActive: true,
    path: '/'
  }
};

export const UserOffline: Story = {
  args: {
    name: 'User',
    avatar: '',
    isActive: false,
    path: '/'
  }
};
