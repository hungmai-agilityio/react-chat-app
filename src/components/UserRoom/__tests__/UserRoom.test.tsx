import { fireEvent, render, screen } from '@testing-library/react';
import { BrowserRouter, MemoryRouter } from 'react-router-dom';

// Components
import { UserRoom } from '@/components/UserRoom';

const propsDefault = {
  id: '9f8a0e26-7761-4e8a-9a64-51ec044fd20f',
  avatar: '',
  name: 'user',
  isActive: true,
  time_stamp: '2024-07-01T02:58:05.791Z',
  path: '/'
};

const onSelectMock = jest.fn();

describe('User room', () => {
  it('renders default item user', () => {
    const { container } = render(
      <MemoryRouter initialEntries={[propsDefault.path]}>
        <UserRoom {...propsDefault} onSelect={onSelectMock} />
      </MemoryRouter>
    );

    const element = screen.getByText(/user/);
    expect(element).toBeInTheDocument();

    const svgElement = container.querySelector('svg');
    expect(svgElement).toBeInTheDocument();
  });

  it('renders item chat with groups', () => {
    const { container } = render(
      <MemoryRouter initialEntries={[propsDefault.path]}>
        <UserRoom
          {...propsDefault}
          isGroup={true}
          message="Hello"
          onSelect={onSelectMock}
        />
      </MemoryRouter>
    );

    const svgElement = container.querySelector('svg');
    expect(svgElement).not.toBeInTheDocument();
  });

  it('should call onSelect with item id when item is clicked', () => {
    render(
      <MemoryRouter initialEntries={[propsDefault.path]}>
        <UserRoom {...propsDefault} onSelect={onSelectMock} />
      </MemoryRouter>
    );

    const itemElement = screen.getByText(/user/);
    fireEvent.click(itemElement);
    expect(onSelectMock).toHaveBeenCalledTimes(1);
  });

  it('should render item with active status', () => {
    const { container } = render(
      <MemoryRouter initialEntries={[propsDefault.path]}>
        <UserRoom {...propsDefault} isActive={true} onSelect={onSelectMock} />
      </MemoryRouter>
    );

    const svgElement = container.querySelector('svg');
    expect(svgElement).toBeInTheDocument();
    expect(svgElement).toHaveClass('text-green-500');
  });

  it('should render item with inactive status', () => {
    const { container } = render(
      <MemoryRouter initialEntries={[propsDefault.path]}>
        <UserRoom {...propsDefault} isActive={false} onSelect={onSelectMock} />
      </MemoryRouter>
    );

    const svgElement = container.querySelector('svg');
    expect(svgElement).toBeInTheDocument();
    expect(svgElement).toHaveClass('text-gray-400');
  });

  it('should be render match to snapshot with group', () => {
    const { container } = render(
      <BrowserRouter>
        <UserRoom
          {...propsDefault}
          isGroup={true}
          message="Hello"
          onSelect={onSelectMock}
        />
      </BrowserRouter>
    );
    expect(container).toMatchSnapshot();
  });

  it('should be render match to snapshot with user', () => {
    const { container } = render(
      <BrowserRouter>
        <UserRoom {...propsDefault} onSelect={onSelectMock} />
      </BrowserRouter>
    );
    expect(container).toMatchSnapshot();
  });
});
