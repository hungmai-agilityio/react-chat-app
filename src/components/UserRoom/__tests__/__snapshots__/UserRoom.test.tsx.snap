// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User room should be render match to snapshot with group 1`] = `
<div>
  <a
    aria-current="page"
    class="flex justify-between items-center w-full h-16 text-sm hover:bg-second px-3 active"
    href="/"
  >
    <div
      class="flex items-center gap-3"
    >
      <div
        class="relative"
      >
        <div
          class="text-primary bg-gray-300 flex items-center justify-center font-semibold rounded-full w-avatar-sm h-avatar-sm text-lg"
          data-testid="avatar"
        >
          U
        </div>
      </div>
      <div>
        <p
          class="font-semibold text-sm w-36 truncate"
        >
          user
        </p>
        <p
          class="w-36 truncate"
        >
          Hello
        </p>
      </div>
    </div>
    <p>
      01.07.2024
    </p>
  </a>
</div>
`;

exports[`User room should be render match to snapshot with user 1`] = `
<div>
  <a
    aria-current="page"
    class="flex justify-between items-center w-full h-16 text-sm hover:bg-second px-3 active"
    href="/"
  >
    <div
      class="flex items-center gap-3"
    >
      <div
        class="relative"
      >
        <div
          class="text-primary bg-gray-300 flex items-center justify-center font-semibold rounded-full w-avatar-sm h-avatar-sm text-lg"
          data-testid="avatar"
        >
          U
        </div>
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-circle absolute text-gray-400 border-xs border-white rounded-full top-6 left-7 text-green-500"
          data-icon="circle"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 512A256 256 0 1 0 256 0a256 256 0 1 0 0 512z"
            fill="currentColor"
          />
        </svg>
      </div>
      <p>
        user
      </p>
    </div>
    <p>
      01.07.2024
    </p>
  </a>
</div>
`;
